### PR TITLE
Migrated all hash functions to Mimcsponge 

### PR DESCRIPTION
**Pull request by kendricktan**
Migrated hash functions to use mimc sponge instead of mimc to fit the specification.

This includes the key derivation. 

Also included some more tests

### CODE REVIEW

File: test/testCircuit.js [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        'message': stringifyBigInts(encryptedMsg),
+        'private_key': stringifyBigInts(sharedKey)
+      }
+
+      const witness = circuit.calculateWitness(circuitInputs)
+
+      // Get the outputs
+      for (let i = 0; i < msg.length; i++) {
+        const idx = circuit.outputIdx(i)
+        assert.equal(msg[i].toString(), witness[idx].toString())
```
Comment (weijiekoh) [inline test/testCircuit.js]: I suggest that we also check if the circuit generates ciphertexts which match the values in `encryptedMsg`.

File: test/testCircuit.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+      const circuitInputs = {
+        'from_x': stringifyBigInts(pk1[0]),
+        'from_y': stringifyBigInts(pk1[1]),
+        'R8x': stringifyBigInts(signature.R8[0]),
+        'R8y': stringifyBigInts(signature.R8[1]),
+        'S': stringifyBigInts(signature.S),
+        'preimage': stringifyBigInts(msg)
+      }
+
+      // If circuit can calculate witness
```
Comment (weijiekoh) [inline test/testCircuit.js]: I suggest adding failing tests which detect exceptions when given invalid signatures.

File: test/testCircuit.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
+const path = require('path')
+const { assert } = require('chai')
+const compiler = require('circom')
+const { stringifyBigInts, unstringifyBigInts } = require('../_build/utils/helpers')
+const { Circuit } = require('snarkjs')
+const { hashLeftRight } = require('../_build/utils/crypto')
```
Comment (weijiekoh) [inline test/testCircuit.js]: I'm not familiar with Flow but I'm not sure that we should import from`_build` rather than from the original source file.

Comment (kendricktan) [reply test/testCircuit.js]: Unfortuntately we mocha doesn't read flow syntax and will throw an error if you import it from 'app' :|

File: test/testCircuit.js (hunk 1 above)
Comment (kendricktan) [reply test/testCircuit.js]: The circuit isn't responsible for encrypting the values, its used to decrypt the encrypted values

**Pull request merged**